nipoppy_new/nipoppy/utils.py

```diff
@@ -0,0 +1,275 @@
+        return json.load(file, **kwargs)
+
+
+def save_json(obj: dict, fpath: str | Path, **kwargs):
+    """Save a JSON object to a file.
+
+    Parameters
+    ----------
+    obj : dict
+        The JSON object
+    fpath : str | Path
+        Path to the JSON file to write
+    indent : int, optional
+        Indentation level, by default 4
+    **kwargs :
+        Keyword arguments to pass to json.dump
+    """
+    if "indent" not in kwargs:
+        kwargs["indent"] = 4
+    fpath = Path(fpath)
+    fpath.parent.mkdir(parents=True, exist_ok=True)
+    with open(fpath, "w") as file:
+        json.dump(obj, file, **kwargs)
+
+
+def add_path_suffix(path: Path | str, suffix: str, sep="-") -> Path:
+    """Add a suffix to a path, before the last file extension (if any)."""
+    path = Path(path)
+    return Path(path.parent, f"{path.stem}{sep}{suffix}{path.suffix}")
+
+
+def add_path_timestamp(
+    path: Path | str, timestamp_format="%Y%m%d_%H%M", sep="-"
+) -> Path:
+    """Add a timestamp to a path, before the last file extension (if any)."""
+    timestamp = datetime.datetime.now().strftime(timestamp_format)
+    return add_path_suffix(path=path, suffix=timestamp, sep=sep)
+
+
+def save_df_with_backup(
+    df: pd.DataFrame,
+    fpath_symlink: str | Path,
+    dname_backups: Optional[str] = None,
+    use_relative_path=True,
+    **kwargs,
+) -> Path | None:
+    """Save a dataframe as a symlink pointing to a timestamped "backup" file.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        The dataframe to save
+    fpath_symlink : str | Path
+        The path to the symlink
+    dname_backups : Optional[str], optional
+        The directory where the timestamped backup file should be written
+        (automatically determined if None), by default None
+    use_relative_path : bool, optional
+        Use relative instead of absolute path for the symlink, by default True
+
+    Returns
+    -------
+    Path or None
+        None if no file was saved, otherwise the path to the backup file
+    """
+    if "index" not in kwargs:
+        kwargs["index"] = False
+
+    fpath_symlink = Path(fpath_symlink)
+
+    fname_backup = add_path_timestamp(fpath_symlink.name)
+    if dname_backups is None:
+        dname_backups = f".{fpath_symlink.stem}s"
+
+    fpath_backup_full: Path = fpath_symlink.parent / dname_backups / fname_backup
+
+    fpath_backup_full.parent.mkdir(parents=True, exist_ok=True)
+    df.to_csv(fpath_backup_full, **kwargs)
+
+    if use_relative_path:
+        fpath_backup_to_link = os.path.relpath(fpath_backup_full, fpath_symlink.parent)
+    else:
+        fpath_backup_to_link = fpath_backup_full
+
+    if fpath_symlink.is_symlink() or fpath_symlink.exists():
+        fpath_symlink.unlink()
+    fpath_symlink.symlink_to(fpath_backup_to_link)
+
+    return Path(fpath_backup_full)
+
+
+def process_template_str(
+    template_str: str, resolve_paths=True, objs=None, lower=True, **kwargs
```

Comment:
A sequence of anything that has attributes! Practically I have been using Nipoppy `Layout` objects and/or workflow objects. This is mostly so that we can have template strings like `[[NIPOPPY_XXX]]` where `XXX` is some attribute of a layout/workflow object (useful for paths in Boutiques invocations).

I will do a pass of docstring writing soon, so there will be documentation of public functions/methods and their parameters.